kolibri/core/device/management/commands/provisiondevice.py

```diff
@@ -0,0 +1,118 @@
+    return answer.lower()
+
+
+languages = dict(settings.LANGUAGES)
+
+
+def create_facility(options):
+    if options['facility'] is None and options['interactive']:
+        answer = get_user_response("Do you wish to create a facility? [yn] ", ["y", "n"])
+        if answer == "y":
+            options['facility'] = get_user_response("What do you wish to name your facility? ")
+
+    if options['facility']:
+        facility, created = Facility.objects.get_or_create(name=options['facility'])
+
+        if not created:
+            logging.warn("Facility with name {name} already exists, not modifying preset.".format(name=options['facility']))
+            return facility
+
+        logging.info("Facility with name {name} created.".format(name=options['facility']))
+
+        if options['preset'] is None and options['interactive']:
+            options['preset'] = get_user_response(
+                "Which preset do you wish to use? [{presets}]: ".format(presets=",".join(presets.keys())),
+                valid_answers=presets)
+
+        # Only set preset data if we have created the facility, otherwise leave previous data intact
+        if options['preset']:
+            preset = options['preset']
+            dataset_data = mappings[preset]
+            for key, value in dataset_data.items():
+                setattr(facility.dataset, key, value)
+            facility.dataset.save()
+            logging.info("Facility preset changed to {preset}.".format(preset=options['preset']))
+    else:
+        facility = Facility.get_default_facility() or Facility.objects.first()
+        if not facility:
+            raise CommandError('No facility exists')
+    return facility
+
+
+def create_superuser(options):
+    if options['superusername'] is None and options['interactive']:
+        options['superusername'] = get_user_response("Enter a username for the super user: ")
+
+    if options['superuserpassword'] is None and options['interactive']:
+        confirm = ""
+        while options['superuserpassword'] != confirm:
+            options['superuserpassword'] = get_user_response("Enter a password for the super user: ")
+            confirm = get_user_response("Confirm password for the super user: ")
+
+    if options['superusername'] and options['superuserpassword']:
+        superuser, created = FacilityUser.objects.get_or_create(username=options['superusername'], facility=options['facility'])
+        if created:
+            logging.info("Superuser created with username {username}.".format(username=options['superusername']))
+            superuser.set_password(options["superuserpassword"])
+            superuser.save()
+        else:
+            logging.info("Setting user with username {username} as superuser.".format(username=options['superusername']))
+        options['facility'].add_role(superuser, ADMIN)
+        DevicePermissions.objects.create(user=superuser, is_superuser=True)
+
+
+def create_device_settings(options):
```

Comment:
The `options` argument of all these utility-like functions is being a bit abused IMO... would be better to handle the command line UI (yes, it's a UI, did you know that? :smile:  ) in the `Command` class. Then set some clear `kwarg` for the utility functions and move them somewhere else.

Advantage:

1. Easier to test the utility functions
1. Other callers may implement them, rather than duplicating
1. Keep Model access etc. in one, unified place
1. Messy CLI code is isolated

Reply:
So this function would be `create_device_settings(language_id=None, facility=None)` and refactored to `core.device.utils` (or `core.device.shortcuts`) and methods `Command` would handle asking the user questions.